cpp/src/groupby/hash/groupby.cu

```diff
@@ -351,40 +414,119 @@ class hash_compound_agg_finalizer final : public cudf::detail::aggregation_final
 };
-// flatten aggs to filter in single pass aggs
-std::tuple<table_view, std::vector<aggregation::Kind>, std::vector<std::unique_ptr<aggregation>>>
-flatten_single_pass_aggs(host_span<aggregation_request const> requests)
-{
-  std::vector<column_view> columns;
-  std::vector<std::unique_ptr<aggregation>> aggs;
-  std::vector<aggregation::Kind> agg_kinds;
 
-  for (auto const& request : requests) {
-    auto const& agg_v = request.aggregations;
+/**
+ * @brief "Flattens" groupby aggregations into its constituent simple aggregations.
+ *
+ * This maintains the mapping between each simple aggregation, the original aggregation column,
+ * and any substituted aggregation column.
+ * E.g.
+ *   1. SUM is already a simple aggregation, and applies directly to the original column.
+ *      No substitution is needed.
+ *      Mapping: [ SUM -> original_agg_col ]
+ *   2. MEAN is a compound aggregation, and is composed of `SUM` and `COUNT`, both applied
+ *      on the original column.
+ *      Mapping: [ SUM -> original_agg_col, COUNT -> original_agg_col ]
+ *   3. NTH_ELEMENT(0, null_policy::INCLUDE) is a more complex aggregation that can be
+ *      expressed in terms of a `MIN` aggregation on a newly generated "source-index" column.
+ *      Mapping: [ MIN -> source_index_col ]
+ */
+class single_pass_aggs_flattener {
+ public:
+  explicit single_pass_aggs_flattener(rmm::cuda_stream_view stream) : _stream{stream} {}
 
-    std::unordered_set<aggregation::Kind> agg_kinds_set;
-    auto insert_agg = [&](column_view const& request_values, std::unique_ptr<aggregation>&& agg) {
-      if (agg_kinds_set.insert(agg->kind).second) {
-        agg_kinds.push_back(agg->kind);
-        aggs.push_back(std::move(agg));
-        columns.push_back(request_values);
-      }
-    };
+  struct result {
+    std::vector<std::unique_ptr<aggregation>> simple_aggs;  ///< List of simple aggregations.
+    std::vector<column_view> agg_columns;           ///< Possibly substituted aggregation columns.
+    std::vector<column_view> original_agg_columns;  ///< Original aggregation columns.
+    std::vector<aggregation::Kind> agg_kinds;       ///< List of simple aggregation kinds.
+    std::vector<std::unique_ptr<column>> temp_columns;  ///< Generated temporary columns.
+  };
 
-    auto values_type = cudf::is_dictionary(request.values.type())
-                         ? cudf::dictionary_column_view(request.values).keys().type()
-                         : request.values.type();
-    for (auto&& agg : agg_v) {
-      groupby_simple_aggregations_collector collector;
+  void flatten(aggregation_request const& request)
+  {
+    _agg_kinds_set.clear();
 
-      for (auto& agg_s : agg->get_simple_aggregations(values_type, collector)) {
-        insert_agg(request.values, std::move(agg_s));
-      }
+    std::for_each(request.aggregations.begin(),
+                  request.aggregations.end(),
+                  [&](auto&& compound_agg) { simplify(compound_agg, request.values); });
```

Comment:
Wondering if we can pass by `auto const&` instead of `auto&&` to constrain this a little more. Applies in a few places through here -- I haven't checked closely to see whether this is allowable.

Reply:
I've replaced all instances I introduced with `auto const&` and `auto&`, as appropriate.